netperf_tool/src/client_node.cpp

```diff
@@ -0,0 +1,160 @@
+#include "rclcpp/serialization.hpp"
+#include "rclcpp/serialized_message.hpp"
+
+#include "utils.hpp"
+
+namespace netperf_tool
+{
+
+using namespace std::chrono_literals;
+
+ClientNode::ClientNode(
+  size_t array_size,
+  std::chrono::nanoseconds target_pub_period,
+  const rclcpp::QoS & pub_qos,
+  const rclcpp::NodeOptions & options)
+: Node("client", "netperf_tool", options),
+  array_size_{array_size},
+  target_pub_period_{target_pub_period}
+{
+  pub_ = this->create_publisher<netperf_tool_interfaces::msg::Bytes>("test_topic", pub_qos);
+  // always keep the vector preallocated, to avoid delays when the timer is triggered
+  bytes_.resize(array_size);
+  client_ = this->create_client<netperf_tool_interfaces::srv::GetResults>("get_results");
+}
+
+bool
+ClientNode::wait_for_server(std::chrono::nanoseconds timeout)
+{
+  auto start = std::chrono::system_clock::now();
+  while (std::chrono::system_clock::now() < start + timeout) {
+    if (client_->service_is_ready()) {
+      return true;
+    }
+    RCLCPP_INFO(this->get_logger(), "perf server not available yet, waiting...");
+    std::this_thread::sleep_for(500ms);
+  }
+  RCLCPP_INFO_STREAM(
+    this->get_logger(),
+    "perf server not available after waiting ["
+      << std::chrono::duration_cast<std::chrono::seconds>(timeout).count() << "s]" << std::endl);
+  return false;
+}
+
+void
+ClientNode::start_publishing()
+{
+  timer_ = this->create_wall_timer(
+    target_pub_period_,
+    std::bind(&ClientNode::pub_next_msg, this));
+}
+
+void
+ClientNode::stop_publishing()
+{
+  if (timer_) {
+    timer_->cancel();
+  }
+}
+
+void
+ClientNode::sync_with_server(rclcpp::Executor & exec)
+{
+  if (!pub_->wait_for_all_acked(5s)) {
+    RCLCPP_WARN(this->get_logger(), "Some messages were not acked by the perf server ...");
+  }
+  if (!client_->service_is_ready()) {
+    RCLCPP_ERROR(
+      this->get_logger(), "Perf server is not available anymore, cannot get the results back");
+    return;
+  }
+  auto req = std::make_shared<netperf_tool_interfaces::srv::GetResults::Request>();
+  req->publisher_gid = this->get_stringified_pub_gid();
+  req->messages_total = this->collected_info_.message_ids.size();
+  auto future = client_->async_send_request(req);
+  exec.spin_until_future_complete(future);
+  statistics_ = future.get();
+}
+
+ClientResults
+ClientNode::extract_results()
+{
+  ClientResults results;
+  if (!statistics_) {
+    throw std::runtime_error{"you must first completely run the client before getting results"};
+  }
+  results.statistics = *statistics_;
+  statistics_.reset();
+  {
+    std::lock_guard guard{collected_info_mutex_};
+    results.collected_info = std::move(collected_info_);
+  }
+  return results;
+}
+
+std::string
+ClientNode::get_topic_name()
+{
+  return pub_->get_topic_name();
+}
+
+std::string
+ClientNode::get_stringified_pub_gid()
+{
+  return stringify_gid(pub_->get_gid());
+}
+
+void
+ClientNode::pub_next_msg()
+{
+  netperf_tool_interfaces::msg::Bytes msg;
```

Comment:
Could you not reuse this message loop-to-loop? You'd be able to reuse the serialized size too, since the id and timestamp shouldn't change size (afaik).

Reply:
> Could you not reuse this message loop-to-loop?

Yes, I will try that.

> You'd be able to reuse the serialized size too, since the id and timestamp shouldn't change size (afaik).

Agreed.
I think I can try to calculate the serialized message size once in the constructor, and then I will not need to deal with serialization when publishing.
I will try that.

Reply:
Done in https://github.com/ros2/ros2cli/pull/720/commits/4cb873ac81d49b34edb5034efc32911fd20bc9ab.